Makefile

```diff
@@ -452,11 +452,19 @@ ifeq ($(GOOS),linux)
 
 include tools/istio-docker.mk
 
-endif # end of docker block that's restricted to Linux
+# if first part of URL (i.e., hostname) is gcr.io then upload istioctl and deb
+$(if $(findstring gcr.io,$(firstword $(subst /, ,$(HUB)))),$(eval push: gcs.push.istioctl-all gcs.push.deb),)
```

Comment:
why is the upload conditional ? i'm a bit confused by this

Reply:
This is based on the original bin/push script, which had something like:

<pre># Permit $HUB overrides to suppress errors during push
if [[ "${HUB}" =~ ^gcr\.io ]]; then
    pilot/bin/upload-istioctl -p "gs://istio-artifacts/pilot/${TAG}/artifacts/istioctl" &
    pilot/bin/push-debian.sh -c opt -p "gs://istio-artifacts/pilot/${TAG}/artifacts/debs" &
    security/bin/push-debian.sh -c opt -p "gs://istio-artifacts/auth/${TAG}/artifacts/debs" &
fi
</pre>

The rule is trying to do the make equivalent, whereby a "push" to GCS is done if the hub starts with gcr.io.  I'm not a fan of the behavior in the original script (not porting it to make) -- it seems like too much of a kludge to assume that a push to GCR also means they want a copy to GCS.  I hesitantly ported the functionality to make so that my deletion of bin/push would be less controversial (by retaining feature-parity, though at the time I did the port I didn't know who/what needed the functionality and only the "upload-istioctl" line still existed [the debian lines were presumably removed as part of the consolidation to one debian file]).

My preference would be to have users do an explicit copy for themselves, thought I don't quite know what that should look like.  Offhand I'm thinking that someone would do a "make push", then a "make artifacts" (or maybe "artifacts" would already be a superset of "push"), and then manually copy the artifacts as needed, e.g.:

<pre>
make push
make artifacts
gsutil -m cp -r "$(make where-is-out)"/istioctl-* "gs://istio-artifacts/pilot/${TAG}/artifacts/debs"
gsutil -m cp -r "$(make where-is-out)"/*.deb "gs://istio-artifacts/pilot/${TAG}/artifacts/debs"
</pre>

and then none of the logic to detect when to do a copy (or defaulting of parameters like "GS_BUCKET") would be needed in the Makefile.

This description isn't necessarily an argument against the change (I did after all suggest it as a way to unblock testing), but I did want to present the background so other people can offer thoughts/opinions on which direction (or offer a third) on if/how the copying of artifacts to GCS should be handled.